### ISSUE TITLE
Any plans to connect luci-go to Google Endpoints v2?

### ISSUE THREAD

**Issue by jimbojetlag** [luci/luci-go#27]
The next version of Google Endpoins in now in beta:

https://cloudplatform.googleblog.com/2016/09/manage-your-APIs-with-Google-Cloud-Endpoints.html

Any plans to make luci-go compatible with this? It seems like the new version works with OpenAPI.

**Comment by riannucci** [comment]
I don't believe that endpoints v2 supports Go at the moment. In any event, we've been having much more success with the [gRPC](http://grpc.io) approach, and have written a lite compatibility layer (https://github.com/luci/luci-go/tree/master/grpc/prpc) to allow us to get a nearly gRPC-compatible API. When gRPC (hopefully) supports AppEngine natively, that prpc layer is designed to be trivially removable so that our apps will work with the native gRPC support.

That said, we tried endpoints v1 and it ended up not working well for our apps (luci-py apps still use endpoints v1). Part of the trouble is that these apps tend to be very procedure-oriented, and REST-spec systems like endpoints v1 and Swagger/OpenAPI lean heavily in the RESTful direction. If we have a service which naturally maps to REST, I think it would definitely be worth consideration (the luci-config service comes to mind). In the mean-time, however, we're seeing very good performance with {p,g}RPC, and the ability to define language neutral APIs (e.g. https://github.com/luci/luci-go/blob/master/logdog/api/endpoints/coordinator/logs/v1/logs.proto#L343) has proven very beneficial.

We've even built a self-hosted RPC explorer (https://luci-logdog.appspot.com/rpcexplorer/services/), as well as a generic command-line utility (https://github.com/luci/luci-go/tree/master/grpc/cmd/rpc) which work with any pRPC app.

**Comment by riannucci** [comment]
Hm... On reading further, it looks like there's some interoperability of gRPC with OpenAPI (https://openapis.org/news/blogs/2016/05/grpc-rest-and-openapi-specification). It may be possible for us to generate OpenAPI documents directly from the protos.

**Comment by riannucci** [comment]
It looks like there's this thing https://github.com/grpc-ecosystem/grpc-gateway which we could potentially use. It looks like the only code requirement in luci-go would be to add `option (google.api.http)` blocks to our service definitions.

This is very different than Endpoints V1. Probably worth further investigation to see what the potential benefits are here.

**Comment by riannucci** [comment]
I think the RESTful bent still applies, however; {g,p}RPC are very RPC oriented, which is pretty different in philosophy than REST. The current pRPC APIs already do have an HTTP+JSON interface though (it's what rpcexplorer and javascript use).

**Comment by jimbojetlag** [comment]
Well, it seems to me that prpc and grpc-gateway have a lot of overlapping: they both try to make grpc available to browser clients. However, grpc-gw converts RPC to json-REST, while prpc converts it to json-RPC.

So this boils down to choosing between REST vs RPC. Each have their own good and bad sides, I'm not discussing that here. What I'd like to point out is that, OpenAPI is on the path to become a de facto, many organizations, including Google, are supporting it. Specifically, according to its PM,  [Google Endpoints v2 also joined OpenAPI](https://news.ycombinator.com/item?id=12407525).

It looks like that if you have an OpenAPI (formerly Swagger) specs for your API, you can connect it to a growing number of managed services. AWS API Gateway already does that, and now Endpoints v2 s following the same path. By this connection, you get a lot of features, like caching, authentication, anti DDoS, etc for 'free'.

If I'm not wrong, allowing prpc to support REST should not require a lot of code:

1. Allow other methods than POST: In addition to `"luci-go/grpc/prpc"Server`, we probably need a `RESTServer` type which [Installhandler](https://github.com/luci/luci-go/blob/master/grpc/prpc/server.go#L149-L158) accepts an http method.
2. Allow other url patterns than `/prpc/:service/:method`: We need a way of 'merging' url query string params into the RPC request instance. At the moment, prpc only cares about the POST body, not the url.

That should do it. To make it even nicer, a swagger doc generator would make luci-go a perfect API development platform in Go. This can be generated by the proto3 `option` keyword, as grpc-gw does.

Having the wagger doc, we should be able tot connect the app to Endpoints v2, it is language neutral and only cares about the swagger specs.

**Comment by riannucci** [comment]
So the only purpose of pRPC is to be as-compatible-as-possible with gRPC (so that when gRPC supports all of our environments the switch will be very painless). As such I don't think we'd want to diverge pRPC very far from where it is today.

That said, I don't think it would be a problem at all to make our apps support a 'really use gRPC' mode apart from pRPC such that they would be usable with grpc-gateway. Then luci-go apps would be bog-standard gRPC components and could be used with grpc-gateway without contortions.
